src/ffi/expression.rs

```diff
@@ -0,0 +1,399 @@
+use crate::ast::{BinaryOperator, Expression, LogicalExpression, Predicate};
+use crate::ffi::ERR_BUF_MAX_LEN;
+use crate::schema::Schema;
+use bitflags::bitflags;
+use std::cmp::min;
+use std::ffi;
+use std::os::raw::c_char;
+use std::slice::from_raw_parts_mut;
+
+impl Expression {
+    fn get_predicates(&self) -> Vec<&Predicate> {
```

Comment:
Not a hard blocker for merge, but should we consider using an `iter()` method instead? I noticed all of the use cases for returned `predicates` are iterations, and an iterator might be a better interface.

Reply:
Will do before this PR getting merged. Iteration interface will improve a little performance.

Reply:
Done. Now use a stack-based iterator. Although using Morris traversal could achieve this with more space efficiency (O(1) in space, O(n) in time), it is more intrusive on the tree structure itself and may require a raw pointer on the tree. I don't think it's worth it.